src/middleware/handle.js

```diff
@@ -31,7 +31,7 @@ const callHandler = ({entry, context, monitor}) => {
       } else if (err.code === 'AuthenticationError') {
         return res.reportError('AuthenticationFailed', err.message, err.details);
       }
-      return res.reportInternalError(err);
```

Comment:
I suspect this is the issue behind the test failures.  You can probably find more by running a single test with `DEBUG=*`.  For example:

```shell
DEBUG=* mocha test/auth_test.js --grep 'function that still uses satisfies fails'
```

Reply:
I am kinda stuck at this. Tried running alot of tests. I tried adding this here
```
if (err.status) {
        return res.status(err.status).json(err);
      }
      return res.reportInternalError(err);
```
but that also doesn't seems to work. 
Can you give me some info on where exactly should I look for?

Reply:
I think this is a case for tracing through what happens by hand.  Let's take, for example, code in an API handler that calls
```js
res.reportError(
  'TooManyFoos',
  'You can only have 3 foos.  These foos already exist:\n```\n{{foos}}\n```',
  {foos: foomanager.foos(request.fooId)});
```

That will throw (from `errors.js`):
```js
       throw ({status, code, message, requestInfo}); 
```

That will be caught (in `handle.js`), and since `err.code` is not `AuthorizationError` or `AuthenticationError`, will execute:
```js
return res.status(err.status).json(err);
```
which is close to correct, although it contains an extra `status` property in the response body that should not be present.

Here are a few more scenarios to trace through:
 * some other exception occurs in the API handler (for example, `throw new Error('oh no!')`)
 * something calls `res.reportError('AuthorizationError', ..)` (this code is handled specially in `handle.js`), or similarly `AuthenticationError`

Once you've done that, what needs to be done should be a little clearer.

Reply:
I did some debugging, got 1 doubt. 
For this test
`https://github.com/taskcluster/taskcluster-lib-api/blob/master/test/errors_test.js#L142`
in this test, throw in `errors.js` is not going to `handle.js`.
Is all the request does not goes to `handle.js` due to which correct request is not returned and this test has failed.

Reply:
This makes sense.

I found one more test which is giving similar issue.
https://github.com/taskcluster/taskcluster-lib-api/blob/master/test/errors_test.js#L67
The test at above line,
it throws the error ` Error('oh no!')` so for building the 500 response it needs to go to `reportInternalError` which is then calling `res.reportError` and when it throws from `reportError` it crashes as `reportInternalError` does not have a catch block.
So I added a try catch block wrapping `reportInternalError` code.
Doing the above change, for separating "reporting an error" and "replying with an error" can be used for this case also, that is, we can call "reporting an error" from `reportInternalError` catch block.

Any thoughts on this? Does this approach makes sense? Or any inputs from your end?

Reply:
Let's see some code :)